apps/server/lib/lexical/server/project/progress/value.ex

```diff
@@ -0,0 +1,39 @@
+defmodule Lexical.Server.Project.Progress.Value do
+  alias Lexical.Protocol.Notifications
+  alias Lexical.Protocol.Types.WorkDone
+
+  @enforce_keys [:token, :kind]
+  defstruct [:token, :kind, :title, :message]
+
+  def begin(title) do
+    token = System.unique_integer([:positive])
+    %__MODULE__{token: token, kind: :begin, title: title}
+  end
+
+  def report(%__MODULE__{token: token}, message) do
+    %__MODULE__{token: token, kind: :report, message: message}
+  end
+
+  def report(_, _) do
+    nil
+  end
+
+  def complete(%__MODULE__{token: token}, message) do
+    %__MODULE__{token: token, kind: :end, message: message}
+  end
+
+  def to_progress(%__MODULE__{kind: :report = kind} = value) do
+    p_value = WorkDone.Progress.Report.new(kind: kind, message: value.message)
```

Comment:
i'm not a big fan of short prefixes like `p_` I'm ok with:
```elixir
Notifications.Progress.new(token: value.token,
  value: WorkDone.Progress.Report.new(kind: value.kind, message: value.message)
  )
  ```
Also, the `kind: :report = kind` is kinda clunky, you don't need to capture the kind, just use value.kind.